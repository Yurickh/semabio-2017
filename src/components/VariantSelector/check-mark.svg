<?xml version="1.0" encoding="UTF-8"?>
<svg width="16px" height="10px" viewBox="0 0 16 10" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 45.2 (43514) - http://www.bohemiancoding.com/sketch -->
    <title>Shape</title>
    <desc>Created with Sketch.</desc>
    <defs></defs>
    <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Artboard-2" transform="translate(-676.000000, -187.000000)" fill-rule="nonzero" fill="#FFFFFF">
            <g id="Group-5" transform="translate(670.000000, 178.000000)">
                <g id="noun_984870_cc" transform="translate(6.000000, 9.000000)">
                    <path d="M15.7628032,1.33934768 L6.96495957,9.83344899 C6.73495058,10.055517 6.38993711,10.055517 6.15992812,9.83344899 L0.237196765,4.11519778 C-0.0790655885,3.80985427 -0.0790655885,3.31020125 0.237196765,3.00485774 C0.553459119,2.69951423 1.07097934,2.69951423 1.38724169,3.00485774 L6.56244385,8.00138793 L14.6127583,0.229007634 C14.9290207,-0.0763358779 15.4465409,-0.0763358779 15.7628032,0.229007634 C16.0790656,0.534351145 16.0790656,1.03400416 15.7628032,1.33934768 Z" id="Shape"></path>
                </g>
            </g>
        </g>
    </g>
</svg>