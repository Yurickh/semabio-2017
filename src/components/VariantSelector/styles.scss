@import 'ambience/color';
@import 'ambience/responsive';

.variant-selector {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;

  @include small-only {
    flex-direction: column;
  }

  & > .variant {
    display: flex;
    align-items: center;
    justify-content: space-between;

    box-sizing: border-box;
    width: 47.5%;
    margin-bottom: 2.5%;
    padding: 26px 23px;

    cursor: pointer;

    border: 1px solid $color-gallery;
    background-color: #FFF;

    @include small-only {
      width: 100%;
      line-height: 1.3;
    }
  }

  &.-shirt > .variant {
    justify-content: center;

    width: 19.2%;

    @include small-only {
      width: 100%;
    }
  }

  & > .variant.-disabled {
    cursor: not-allowed;

    opacity: .4;
    background-color: $color-gallery;

    &::after {
      display: block;

      content: 'esgotado';
      text-transform: uppercase;

      color: red;
    }
  }

  & > .variant.-selected {
    position: relative;

    box-shadow: rgba($color-description, .5) 0 2px 24px;

    & > .tick {
      position: absolute;
      top: -7px;
      right: -10px;

      display: flex;
      align-items: center;
      justify-content: center;

      width: 28px;
      height: 28px;

      border-radius: 50%;
    }
  }
}
