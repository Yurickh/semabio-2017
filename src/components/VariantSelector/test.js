import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'

import VariantSelector from '../VariantSelector'

describe('VariantSelector', () => {
  it('should render correctly', () => {
    const tree = shallow(<VariantSelector />)
    expect(toJSON(tree)).toMatchSnapshot()
  })
})
