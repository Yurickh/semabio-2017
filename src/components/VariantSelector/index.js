import React, { Component } from 'react'
import PropTypes from 'prop-types'

import ShopClient from '../../helpers/shopClient'
import GraphQL from '../../helpers/GraphQL'

import checkMark from './check-mark.svg'

import './styles.css'

class VariantSelector extends Component {
	state = {
		variants: [],
		selected: [],
	}

	static propTypes = {
		product: PropTypes.oneOf(['A', 'B', 'C', 'D', 'E', 'shirt']).isRequired,
		currentColor: PropTypes.string.isRequired,
		onChange: PropTypes.func,
		onLoad: PropTypes.func,
	}

	componentDidMount() {
		this.shopClient = ShopClient()

		this.componentDidUpdate({})
	}

	componentDidUpdate(prevProps) {
		if (this.props.product !== prevProps.product) {
			this.fetchProduct()
		}
	}

	fetchProduct = () => {
		const { product } = this.props

		this.shopClient.fetchQueryProducts({ title: this.nameFor(product)})
		.then(data => {
			const productData = GraphQL.read(data)

			this.setState({
				variants: productData.first().get('variants')
					.map(variant => ({
						id: variant.get('id'),
						name: variant.get('title'),
						available: variant.get('available')
					}))
			}, this.props.onLoad)
		})
	}

	nameFor = (product) => {
		if (product.length === 1) {
			return `Bloco ${product}`
		} else if (this.props.package) {
			return `Pacote ${this.props.package}`
		} else {
			return 'Camiseta'
		}
	}

	allowedToSelect = () => {
		const { variants } = this.state

		return this.props.allowedToSelect && variants.filter(this.variantIsSelected).length === 0
	}

	selectVariant = (variant) => {
		if (!this.allowedToSelect()) return

		this.setState({
			selected: this.state.selected.concat(variant)
		}, () => this.props.onChange(this.state.selected))
	}

	removeVariant = (variant) => {
		this.setState({
			selected: this.state.selected.filter(x => x.id !== variant.id)
		}, () => this.props.onChange(this.state.selected))
	}

	toggleVariant = (variant) => {
		if (this.variantIsSelected(variant)) {
			this.removeVariant(variant)
		} else {
			this.selectVariant(variant)
		}
	}

	variantIsSelected = (variant) => {
		const { selected } = this.state

		return !!selected.filter(selection => selection.id === variant.id).length
	}

	variant = (variant) => {
		if (!variant.available) {
			return '-disabled'
		} else if (this.variantIsSelected(variant)) {
			return '-selected'
		} else {
			return ''
		}
	}

	isShirtSelector = () => {
		return !!this.props.package
	}

	selectorVariant = () => {
		return this.isShirtSelector() ? '-shirt' : ''
	}

	variantName = (variant) => {
		if (!this.isShirtSelector()) return variant.name

		return variant.name.replace('Camiseta (', '').replace(')', '')
	}

	render() {
		return (
			<div className={`variant-selector ${this.selectorVariant()}`}>
				{
					this.state.variants
						.map((variant, index) => (
							<div key={index}
								className={`variant ${this.variant(variant)}`}
								onClick={() => this.toggleVariant(variant)}
							>
								{this.variantName(variant)}
								{this.variantIsSelected(variant) && (
									<div className="tick" style={{ backgroundColor: this.props.currentColor }}>
										<img src={checkMark} alt="check mark" />
									</div>
								)}
							</div>
						))
				}
			</div>
		)
	}
}

export default VariantSelector
