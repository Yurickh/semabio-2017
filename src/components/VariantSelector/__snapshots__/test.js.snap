// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariantSelector should render correctly 1`] = `
<div
  className="variant-selector "
/>
`;
